File: 백준/Gold/14500. 테트로미노/테트로미노.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, M] = input[0].split(' ').map(Number);
const board = input.slice(1).map(line => line.split(' ').map(Number));

const dx = [0, 0, 1, -1];
const dy = [1, -1, 0, 0];

let visited = Array.from({ length: N }, () => Array(M).fill(false));
let maxSum = 0;

function dfs(x, y, depth, sum) {
  if (depth === 4) {
    maxSum = Math.max(maxSum, sum);
    return;
  }

  for (let d = 0; d < 4; d++) {
    const nx = x + dx[d];
    const ny = y + dy[d];
    if (nx >= 0 && ny >= 0 && nx < N && ny < M && !visited[nx][ny]) {
      visited[nx][ny] = true;
      dfs(nx, ny, depth + 1, sum + board[nx][ny]);
      visited[nx][ny] = false;
    }
  }
}

// ㅗ, ㅓ, ㅏ, ㅜ 따로 처리
function checkExtraShape(x, y) {
  const shapes = [
    // ㅗ
    [[0, 0], [0, -1], [0, 1], [-1, 0]],
    // ㅜ
    [[0, 0], [0, -1], [0, 1], [1, 0]],
    // ㅓ
    [[0, 0], [-1, 0], [1, 0], [0, -1]],
    // ㅏ
    [[0, 0], [-1, 0], [1, 0], [0, 1]],
  ];

  for (const shape of shapes) {
    let sum = 0;
    let isValid = true;

    for (const [dx, dy] of shape) {
      const nx = x + dx;
      const ny = y + dy;
      if (nx >= 0 && ny >= 0 && nx < N && ny < M) {
        sum += board[nx][ny];
      } else {
        isValid = false;
        break;
      }
    }

    if (isValid) {
      maxSum = Math.max(maxSum, sum);
    }
  }
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    visited[i][j] = true;
    dfs(i, j, 1, board[i][j]);
    visited[i][j] = false;
    checkExtraShape(i, j); // ㅗ는 DFS로만들 수 없는 모양임 -> 방문한 위치에 다시 연속으로 만들 수 없기때문에
    // 남은 부분에 예외모양으로 채워넣기임
  }
}

console.log(maxSum);

File: 백준/Gold/14500. 테트로미노/테트로미노.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	s      = bufio.NewScanner(os.Stdin)
	w      = bufio.NewWriter(os.Stdout)
	n, m   int
	maxSum int

	visited [][]bool
	board   [][]int
	dx      = []int{0, 0, -1, 1}
	dy      = []int{-1, 1, 0, 0}
)

func main() {

	defer w.Flush()
	s.Split(bufio.ScanLines)

	hw := scanLines()
	n, m = hw[0], hw[1]

	board = make([][]int, n)
	visited = make([][]bool, n)

	for i := 0; i < n; i++ {
		visited[i] = make([]bool, m)
		line := scanLines()
		board[i] = line
	}

	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			visited[i][j] = true
			bt(i, j, 1, board[i][j])
			visited[i][j] = false
			extraShapes(i, j)
		}
	}

	fmt.Fprintln(w, maxSum)
}

func bt(x, y, depth, sum int) {
	if depth == 4 {
		if maxSum < sum {
			maxSum = sum
		}
		return
	}

	for i := 0; i < 4; i++ {
		nx, ny := x+dx[i], y+dy[i]
		if nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] {
			visited[nx][ny] = true
			bt(nx, ny, depth+1, sum+board[nx][ny])
			visited[nx][ny] = false
		}
	}
}

func extraShapes(x, y int) {

	shapes := [][][]int{
		{{0, 0}, {-1, 0}, {1, 0}, {0, -1}},
		{{0, 0}, {-1, 0}, {1, 0}, {0, 1}},
		{{0, 0}, {0, 1}, {1, 0}, {0, -1}},
		{{0, 0}, {0, 1}, {-1, 0}, {0, -1}},
	}

	for _, shape := range shapes {
		sum := 0
		founded := true

		for _, v := range shape {
			nx, ny := x+v[0], y+v[1]
			if nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] {
				sum += board[nx][ny]
			} else {
				founded = false
				break
			}

		}

		if founded {
			if maxSum < sum {
				maxSum = sum
			}
		}

	}

}

func scanLines() []int {
	s.Scan()
	parts := strings.Fields(s.Text())
	num := make([]int, len(parts))

	for i, v := range parts {
		num[i], _ = strconv.Atoi(v)
	}

	return num
}
